torchrl/data/rlhf/utils.py

```diff
@@ -0,0 +1,293 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+from typing import Tuple
+
+import torch
+
+from tensordict import TensorDict
+from torch import Tensor
+from torch.nn import functional as F
+
+from torchrl.data.rlhf.prompt import PromptData
+from transformers import GenerationConfig
+
+
+class RolloutFromModel:
+    """A class for performing rollouts with language models.
+
+    Args:
+        model (transformers.Transformer): the model to be used. Should have a
+            :meth:`generate` method.
+        ref_model (transformers.Transformer): a frozen version of ``model``
+            where params are in their initial configuration.
+        reward_model: (nn.Module, tensordict.nn.TensorDictModule): a model which, given
+            input_ids and attention_mask, calculates rewards for each token and
+            end_scores (the reward for the final token in each sequence).
+        max_new_tokens (int, optional): the maximum length of the sequence.
+            Defaults to 50.
+    """
+
+    EOS_TOKEN_ID = 50256
+
+    def __init__(self, model, ref_model, reward_model, max_new_tokens=50):
+        self.model = model
+        self.ref_model = ref_model
+        self.reward_model = reward_model
+        self.max_new_tokens = max_new_tokens
+
+    def kl_step(self):
```

Comment:
why we have both kl_step and the rollout functions take kl_coef as input?
isn't choosing either of the two approaches enough?

Reply:
Yes it is, we should move the KL scheduler here, but currently that is not defined on this branch hence the placeholder.